### PR TITLE
recursive-patterns(11): use stack spiller for switch expressions

### PR DESCRIPTION
**Pull request by gafter**
This refactoring moves the spiller between initial lowering and lambda lowering
- Added some improved debug-time display for synthesized locals
- Added some invariant checking regarding bound nodes.
- Move invocation of the spiller into LocalRewriter.Rewrite
- Add an error for when an expression tree contains a switch expression
- Made the local substituter apply at the top level in the expression spiller because expressions can now contain statements (BoundSpillSequence)
- Move the error ERR_ByRefTypeAndAwait to IteratorAndAsyncCaptureWalker

There are two iterations:
- Most of the changes
- Renaming and moving a file

I separated them to make review easier, as the rename and move makes it look like a file was deleted and one added.

This is built on top of other pending reviews, so even though there are only two new iteration here, you will see iterations from previous PRs. Only the last two matter for this PR.

@agocke @cston Can you review this please?
/cc @dotnet/roslyn-compiler

### CODE REVIEW

**Comment by gafter** [comment]
This refactoring moves the spiller between initial lowering and lambda lowering
- Added some improved debug-time display for synthesized locals
- Added some invariant checking regarding bound nodes.
- Move invocation of the spiller into LocalRewriter.Rewrite
- Add an error for when an expression tree contains a switch expression
- Made the local substituter apply at the top level in the expression spiller because expressions can now contain statements (BoundSpillSequence)
- Move the error ERR_ByRefTypeAndAwait to IteratorAndAsyncCaptureWalker

There are two iterations:
- Most of the changes
- Renaming and moving a file

I separated them to make review easier, as the rename and move makes it look like a file was deleted and one added.

This is built on top of other pending reviews, so even though there are only two new iteration here, you will see iterations from previous PRs. Only the last two matter for this PR.

@agocke @cston Can you review this please?
/cc @dotnet/roslyn-compiler

**Comment by gafter** [comment]
@dotnet/roslyn-interactive I had to skip about 50 interactive tests that are producing bound trees that violate bound tree invariants (they use locals that are not declared). Please see https://github.com/dotnet/roslyn/issues/25702 for details.

**Comment by cston** [comment]
> using System;

Copyright.

---
Refers to: src/Compilers/Core/Portable/Collections/IOrderedReadOnlySet.cs:1 in f7b666d. [](commit_id = f7b666db04d060a650c9550d13e1d13acd28b1ae, deletion_comment = False)

**Comment by cston** [comment]
> using System;

Alternatively, perhaps change `MethodToStateMachineRewriter._hoistedVariables` to `OrderedSet<Symbol>` and remove this interface.

---
In reply to: [376626943](https://github.com/dotnet/roslyn/pull/25697#issuecomment-376626943) [](ancestors = 376626943)

---
Refers to: src/Compilers/Core/Portable/Collections/IOrderedReadOnlySet.cs:1 in f7b666d. [](commit_id = f7b666db04d060a650c9550d13e1d13acd28b1ae, deletion_comment = False)

File: src/Compilers/Core/Portable/Symbols/RefKind.cs [hunk 1]
```diff
@@ -76,9 +76,11 @@ internal static string ToParameterPrefix(this RefKind kind)
             }
         }
 
-        // used internally to track `In` arguments that were specified with `In` modifier
-        // as opposed to those that were specified with no modifiers and matched `In` parameter
-        // There is at least one kind of anlysis that cares about this distinction - async stack spilling
+        // Used internally to track `In` arguments that were specified with `In` modifier
+        // as opposed to those that were specified with no modifiers and matched `In` parameter.
+        // There is at least one kind of anlysis that cares about this distinction - hoisting
```
Comment (cston) [inline src/Compilers/Core/Portable/Symbols/RefKind.cs]: >anlysis [](start = 41, length = 7)

Typo.

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 2]
```diff
@@ -5348,4 +5348,7 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="WRN_SwitchExpressionNotExhaustive_Title" xml:space="preserve">
     <value>The switch expression does not handle all possible inputs (it is not exhaustive).</value>
   </data>
+  <data name="ERR_ExpressionTreeContainsSwitchExpression" xml:space="preserve">
+    <value>An expression tree may not contain a switch-expression.</value>
```
Comment (cston) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: >switch-expression [](start = 48, length = 17)

Perhaps `switch expression` (no hyphen) to match the other diagnostics.

File: src/Compilers/CSharp/Portable/BoundTree/BoundNode.cs [hunk 3]
```diff
@@ -172,5 +194,227 @@ internal string GetDebuggerDisplay()
             }
             return result;
         }
+
+        public void CheckLocalsDefined()
```
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNode.cs]: >CheckLocalsDefined [](start = 20, length = 18)

`[Conditional("DEBUG")]`

File: src/Compilers/CSharp/Portable/BoundTree/BoundNode.cs [hunk 4]
```diff
@@ -172,5 +194,227 @@ internal string GetDebuggerDisplay()
...
+            }
+
+            public static void CheckLocalsDefined(BoundNode root)
+            {
+                var localsScanner = new LocalsScanner();
+                localsScanner.Visit(root);
+                localsScanner.Free();
+            }
+
+            void AddAll(ImmutableArray<LocalSymbol> locals)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNode.cs]: >void [](start = 12, length = 4)

`private` for these three methods.

File: src/Compilers/CSharp/Portable/Lowering/DiagnosticsPass_ExpressionTrees.cs [hunk 5]
```diff
@@ -96,6 +96,16 @@ public override BoundNode VisitBaseReference(BoundBaseReference node)
             return base.VisitBaseReference(node);
         }
 
+        public override BoundNode VisitSwitchExpression(BoundSwitchExpression node)
+        {
+            if (_inExpressionLambda)
+            {
+                Error(ErrorCode.ERR_ExpressionTreeContainsSwitchExpression, node);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/DiagnosticsPass_ExpressionTrees.cs]: >ERR_ExpressionTreeContainsSwitchExpression [](start = 32, length = 42)

Is there a test for this code path?

**Comment by cston** [comment]
>                     case BoundKind.Call:

Where did this code move to?

---
Refers to: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AwaitExpressionSpiller.cs:379 in 7062ed2. [](commit_id = 7062ed2ec54c6ef854bb8be4a0dfa6f869e98fba, deletion_comment = True)

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AwaitExpressionSpiller.cs [hunk 6]
```diff
@@ -605,9 +526,23 @@ public override BoundNode DefaultVisit(BoundNode node)
...
-            return builder.Update(replacement);
+            _F.Syntax = node.Syntax;
+            builder.AddStatements(VisitList(node.SideEffects));
+            builder.AddLocals(node.Locals);
+            var value = VisitExpression(ref builder, node.Value);
+            value = Spill(builder, value);
+            return builder.Update(value);
         }
 
         public override BoundNode VisitAddressOfOperator(BoundAddressOfOperator node)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AwaitExpressionSpiller.cs]: >_F.Syntax = node.Syntax; [](start = 12, length = 24)

Why is this necessary? Consider adding a comment.

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AwaitExpressionSpiller.cs [hunk 7]
```diff
@@ -1111,6 +1044,23 @@ public override BoundNode VisitConditionalReceiver(BoundConditionalReceiver node
...
+            _F.CurrentMethod = node.Symbol;
+            var result = base.VisitLambda(node);
+            _F.CurrentMethod = oldCurrentMethod;
+            return result;
+        }
+
+        public override BoundNode VisitLocalFunctionStatement(BoundLocalFunctionStatement node)
+        {
+            var oldCurrentMethod = _F.CurrentMethod;
+            _F.CurrentMethod = node.Symbol;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AwaitExpressionSpiller.cs]: >VisitLambda [](start = 30, length = 11)

We don't need a separate `SpillSequenceSpiller` instance when visiting lamba and local function bodies?

File: src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/IteratorAndAsyncCaptureWalker.cs [hunk 8]
```diff
@@ -75,10 +75,17 @@ public static OrderedSet<Symbol> Analyze(CSharpCompilation compilation, MethodSy
                     var variable = kvp.Key;
                     var type = (variable.Kind == SymbolKind.Local) ? ((LocalSymbol)variable).Type : ((ParameterSymbol)variable).Type;
 
-                    foreach (CSharpSyntaxNode syntax in kvp.Value)
+                    if (variable is SynthesizedLocal local && local.SynthesizedKind == SynthesizedLocalKind.Spill)
                     {
-                        // CS4013: Instance of type '{0}' cannot be used inside an anonymous function, query expression, iterator block or async method
-                        diagnostics.Add(ErrorCode.ERR_SpecialByRefInLambda, syntax.Location, type);
+                        diagnostics.Add(ErrorCode.ERR_ByRefTypeAndAwait, local.Locations[0], local.Type);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/IteratorAndAsyncCaptureWalker.cs]: >ERR_ByRefTypeAndAwait [](start = 50, length = 21)

Does this correspond to the restricted type case previously in `AwaitExpressionSpiller.AddLocal`? If so, perhaps add `Debug.Assert(local.Type.IsRestrictedType());`.

File: src/Compilers/CSharp/Portable/Syntax/LambdaUtilities.cs [hunk 9]
```diff
@@ -418,6 +418,13 @@ internal static bool IsClosureScope(SyntaxNode node)
                     // that lambda needs a closure that captures the analysis payload of the constructor.
                     return true;
 
+                case SyntaxKind.SwitchExpression:
+                case SyntaxKind.AwaitExpression:
+                    // These translate into a BoundSpillSequence, which is then translated into a block
+                    // containins temps required for spilling subexpressions. That block has the syntax of the switch
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Syntax/LambdaUtilities.cs]: >containins [](start = 23, length = 10)

Typo

File: src/Compilers/CSharp/Portable/BoundTree/BoundDecisionDag.cs [hunk 10]
```diff
@@ -83,44 +92,45 @@ public ImmutableArray<BoundDecisionDag> TopologicallySortedNodes()
...
+            return this.Update(newRoot);
 
             // Make a replacement for a given node, using the precomputed replacements for its successors.
-            BoundDecisionDag makeReplacement(BoundDecisionDag dag)
+            BoundDecisionDagNode makeReplacement(BoundDecisionDagNode dag)
             {
                 switch (dag)
                 {
-                    case BoundEvaluationPoint p:
+                    case BoundEvaluationDecisionDagNode p:
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/BoundTree/BoundDecisionDag.cs]: Are these types sealed?   (for faster typechecks)

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs [hunk 11]
```diff
@@ -92,10 +93,22 @@ internal sealed partial class LocalRewriter : BoundTreeRewriterWithStackGuard
...
                 var loweredStatement = (BoundStatement)localRewriter.Visit(statement);
+                loweredStatement.CheckLocalsDefined();
                 sawLambdas = localRewriter._sawLambdas;
                 sawLocalFunctions = localRewriter._sawLocalFunctions;
                 sawAwaitInExceptionHandler = localRewriter._sawAwaitInExceptionHandler;
+
+                if (localRewriter._needsSpilling && !loweredStatement.HasErrors)
+                {
+                    // Move spill sequences to a top-level statement. This handles "lifting" await and the switch expression.
+                    // PROTOTYPE(patterns2): should this be something that the caller does? It isn't really a "local" rewrite.
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs]: I think it is better be handled by the caller, similarly to handling `await` in exception handling regions. 
- to have clear sequence of tree transformation. 
- to have better attribution when investigating perf impact of lowering passes.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Await.cs [hunk 12]
```diff
@@ -1,15 +1,41 @@
...
+                // Await expression is already at the statement level.
+                return rewrittenAwait;
+            }
+
+            // The await expression will be lowered to code that involves the use of side-effects
+            // such as jumps and labels, therefore it is represented by a BoundSpillSequence.
+            // The resulting nodes will be "spilled" to move such statements to the top
+            // level (i.e. into the enclosing statement list).
+            _needsSpilling = true;
+            BoundLocal replacement = _factory.StoreToTemp(
```
[the remark this answers is not in the document]
Comment (VSadov) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Await.cs]: `await` is never an lvalue, so this temp is never a ref temp. 
In general case we can have byref spills. 

In a case of `await`, they will be later decomposed into something that can be hoisted like {array, index} pairs or will produce an error if it is a true byref spill - like a result of byref method call.

In a case when spill is caused by expression switch or stackalloc (in the future), byref spills would work just fine.

Perhaps need a bit more care to make sure that "readonliness" matches. - I.E. if spilling an `in` argument, the temp must be `ref readonly`. Not sure if this is already handled correctly.

File: src/Compilers/CSharp/Portable/Lowering/SpillSequenceSpiller.cs [hunk 13]
```diff
@@ -14,14 +14,14 @@
 
 namespace Microsoft.CodeAnalysis.CSharp
 {
-    internal sealed class AwaitExpressionSpiller : BoundTreeRewriterWithStackGuard
+    internal sealed class SpillSequenceSpiller : BoundTreeRewriterWithStackGuard
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Lowering/SpillSequenceSpiller.cs]: We are likely unable to handle some expressions that are not permitted in async methods - unsafe stuff and such.

I think it is a separate workitem, but we need to make sure that all or nearly all bound expresions can be spilled.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Await.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (VSadov) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Await.cs]: The part about "readonly" is probably better handled as a separate workitem. It would be a new scenario for spiller. 
In `await` case we decompose ref spills, so readonlyness of temps did not matter since they would not survive.

File: src/Compilers/CSharp/Portable/Lowering/DiagnosticsPass_ExpressionTrees.cs (hunk 5 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Lowering/DiagnosticsPass_ExpressionTrees.cs]: A test will be added in a forthcoming iteration.

---
In reply to: [177542903](https://github.com/dotnet/roslyn/pull/25697#discussion_r177542903) [](ancestors = 177542903)

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AwaitExpressionSpiller.cs (hunk 7 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AwaitExpressionSpiller.cs]: Correct, because things are spilled up to the statement level. Lambdas and local functions have a top-level block containing statements at this stage.

---
In reply to: [177557467](https://github.com/dotnet/roslyn/pull/25697#discussion_r177557467) [](ancestors = 177557467)

File: src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/IteratorAndAsyncCaptureWalker.cs (hunk 8 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/IteratorAndAsyncCaptureWalker.cs]: That is true by construction. There is only one line that adds to the set from which the symbol was fetched:

``` c#
            if (type.IsRestrictedType())
            {
                // ...

                _lazyDisallowedCaptures.Add(variable, syntax);
            }
```


---
In reply to: [177570217](https://github.com/dotnet/roslyn/pull/25697#discussion_r177570217) [](ancestors = 177570217)

File: src/Compilers/CSharp/Portable/BoundTree/BoundDecisionDag.cs (hunk 10 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/BoundTree/BoundDecisionDag.cs]: The bound node generator only generates `abstract` or `sealed` types.

---
In reply to: [178199580](https://github.com/dotnet/roslyn/pull/25697#discussion_r178199580) [](ancestors = 178199580)

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs (hunk 11 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs]: I placed this here because there is more than one caller. However We could duplicate this code for each caller. I am not sure which is better, but this simplifies the job of the caller.

---
In reply to: [178200828](https://github.com/dotnet/roslyn/pull/25697#discussion_r178200828) [](ancestors = 178200828)

File: src/Compilers/CSharp/Portable/Lowering/SpillSequenceSpiller.cs (hunk 13 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Lowering/SpillSequenceSpiller.cs]: The switch expression can appear in an unsafe method, so spilling needs to work with almost everything. If you have specific scenarios that would illustrate the issues, that would be very helpful. Please file as a fresh issue.

---
In reply to: [178203355](https://github.com/dotnet/roslyn/pull/25697#discussion_r178203355) [](ancestors = 178203355)

**Comment by gafter** [comment]
> using System;

`OrderedSet` is mutable. I did this to avoid exposing mutability.

---
In reply to: [376628270](https://github.com/dotnet/roslyn/pull/25697#issuecomment-376628270) [](ancestors = 376628270,376626943)

---
Refers to: src/Compilers/Core/Portable/Collections/IOrderedReadOnlySet.cs:1 in f7b666d. [](commit_id = f7b666db04d060a650c9550d13e1d13acd28b1ae, deletion_comment = False)
